packages/nodejs-ext/scripts/extension.js

```diff
@@ -25,28 +25,60 @@ const EXT_PATH = path.join(__dirname, "/../ext/")
 const testExtensionFailure =
   process.env._TEST_APPSIGNAL_EXTENSION_FAILURE === "true"
 
-function download(url, outputPath) {
-  return new Promise((resolve, reject) => {
-    if (testExtensionFailure) {
-      throw new Error("AppSignal internal test failure")
-    }
+class DownloadError extends Error {
+  constructor(message, downloadUrl) {
+    super(message)
+    this.name = this.constructor.name
+    this.downloadUrl = downloadUrl
+  }
+}
 
+function downloadFromMirror(mirror, filename, outputPath) {
+  return new Promise((resolve, reject) => {
+    const url = path.join(mirror, AGENT_VERSION, filename)
     const file = fs.createWriteStream(outputPath)
 
     https.get(url, response => {
       const { statusCode } = response
 
       if (statusCode >= 400) {
         return reject(
-          new Error(`Request to CDN failed with code HTTP ${statusCode}`)
+          new DownloadError(`Request to CDN failed with code HTTP ${statusCode}`, url)
         )
+      } else {
+        response.pipe(file).on("finish",() => resolve(url))
       }
-
-      response.pipe(file).on("finish", () => resolve(outputPath))
     })
   })
 }
 
+function download(mirrors, filename, outputPath) {
+  return new Promise((resolve, reject) => {
+    if (testExtensionFailure) {
+      throw new DownloadError("AppSignal internal test failure", undefined)
+    }
+
+    if (mirrors.length === 0) {
+      reject(new DownloadError("Could not download agent from any mirror", undefined))
```

Comment:
```suggestion
      reject(new DownloadError("Could not download agent from any mirror", undefined))
      return
```

Turns out we do need a return here. Not to return the promise result, but to have the function exit early. (Calling `reject` doesn't stop the function.) Otherwise, if any error occurs for all mirrors in the `downloadFromMirror` function, it will fall into an infinite loop, retrying an empty mirrors Array over and over and over and over again.

We also could use `throw` here, like we do on line 58, but I probably prefer to use reject. Line 58 should also be updated (in another PR).